### DISCUSSION TITLE
Embarassing Parallelization does not reduce computational time

### DISCUSSION THREAD

**Discussion by Niccolo-Marcucci** [NanoComp/meep#2008] [Q&A]
I need to run many identical (small) simulations in which i need to change few parameters each time. Basically I need something like the "parfor" tool in MATLAB.
I first tried with the Pool() from multiprocessing and then tried with the meep.divide_parallel_processes(N) option in meep (thus using mpirun and the parallel version of meep) but in both cases, for some reason when i run N simulation in parallel, each simulation just takes N times longer to compute.

I'll give an example to be more clear.
- I have a 6 (hardware) cores cpu. 
- I have a simulation that depends on a parameter "x" that i want to test with 6 different value and a parameter "y" which we keep costant for simplicity.  
Then I create a function that defines the simulation, runs it and prints the results to a file. This function takes as input the parameter "x". 
To run in parallel the simulation I'd do the following
```
def simulation_fun(x,y):
  ......
 
x = numpy.linspace(0,1,6)
y = 0
parameter_list = [ (xx, y) for xx in x]
with Pool(6) as parfor:
  parfor.starmap(simulation_fun, parameter_list)
```
or instead 
```
def simulation_fun(x):
  ......
 
x = numpy.linspace(0,1,6)
y = 0
j = meep.divide_parallel_processes(6) 
simulation_fun(x[j], y)
```
Running one simulation at a time takes roughly 100s each, while in both those cases running the parallel version takes 600s (at least in order of magnitude). 
I don't understand how this happens, if I change simulation_fun() with something non meep related it works (at least with Pool()). What can I be doing wrong that forces the parallel computation to slow down when running independent simulations?

P.S. with both methods the results of the simulations are correct.

**Comment by smartalecH** [comment]
> I need to run many identical (small) simulations in which i need to change few parameters each time.

`divide_parallel_processes` should work. Perhaps you have a bug somewhere else in your code?

**Reply by Niccolo-Marcucci** [reply]
I really wish it is so, but the only thing that calls for parallelization in my code is `divide_parallel_processes`, What could be interfering with it?
I use the Conda latest meep version, and i have the same problem both on my machine and in a cluster system.

**Reply by Niccolo-Marcucci** [reply]
@smartalecH I provided a test code in a [reply below](https://github.com/NanoComp/meep/discussions/2008#discussioncomment-2399086), are you still sure this is not a proper issue? Has anyone ever managed to use embarassing parallelization with meep?

**Comment by oskooi** [comment]
What you are seeing could be related to #882 which describes resource contention (particularly memory bandwidth) among multiple simultaneous jobs running on the same machine. One thing you can try is to use Meep compiled with [single-precision floating point for the fields and material arrays](https://meep.readthedocs.io/en/latest/Build_From_Source/#floating-point-precision-of-the-fields-and-materials-arrays) but this would require building Meep from source since the Conda packages are built using double-precision floating point.

**Reply by Niccolo-Marcucci** [reply]
Thanks for the reply. 
What is experienced in that issue seem something in which the computational speed changes with time more or less randomly. Here I get something more systematic i think. Also I should experience the same slowing down when running single simulations parallelized over multiple cores. Which I don't, parallelization over the single simulation works fine (with correct results and speed improvement).

**Reply by oskooi** [reply]
The slowdown reported in #882 is *not* random: it happens whenever a new job is launched while another one is already running. Note that for a given simulation, if you run it as (a) N single-processor simulations compared to (b) a single parallel simulation with N processors, then the *memory per processor* in (b) will be roughly smaller by a factor of 1/N compared to one processor in (a) assuming Meep partitions the cell into equal-sized chunks (the default `split_chunks_evenly=True`). As a result, the memory bandwidth will be larger in (a) than it is in (b). Depending on your particular multicore machine, this could have a significant impact on performance.

**Reply by Niccolo-Marcucci** [reply]
I think I understand what you mean, but I don't get how I could end up with even worst performances using parallelization, especially since there is no communication whatsoever between the jobs. Also the simulations I'm testing (see [my latest reply](https://github.com/NanoComp/meep/discussions/2008#discussioncomment-2399086)) are so small that I'm expecting to have negligible effects on memory bandwidth.

**Comment by stevengj** [comment]
> j = meep.divide_parallel_processes(6) 

Have you checked that `meep.count_processors()` returns 6 if you call it *before* this line?  (To ensure that you are running the correct MPI version of Meep and that you launched it correctly.)

**Reply by Niccolo-Marcucci** [reply]
Thanks for the reply. Yes it does, it returns 6. 
Maybe I should add that parallelization over a single simulation works also fine.

**Comment by Niccolo-Marcucci** [comment]
Ok, let's put it this way. The simulation I run is in [this script](https://github.com/Niccolo-Marcucci/Nanophotonics-meep/blob/master/s08-2D_circular_DBR_cavity.py), but it is uselessly complex for what I'm trying to say. 
Therefore I wrote an overly simplified script that has the exact behaviour i tried to describe. The simulation part of the script comes directly from the tutorial of the bend waveguide, so no error could come from there. You can find the script in my repo: [https://github.com/Niccolo-Marcucci/Nanophotonics-meep/blob/testing/test_embarassing_parallelization.py](https://github.com/Niccolo-Marcucci/Nanophotonics-meep/blob/testing/test_embarassing_parallelization.py)

You can execute the script in 3 ways: serial, parallel with `Poll()` and parallel with `mpirun`
For the first one, when calling the script just include the word `serial` after the script name
```
(pmp) [user@localhost folder]$ python test_embarassing_parallelization.py serial
.......

Elapsed run time = 12.8205 s
```
For the second one, just lunch the script with python (this is what would run if you use spyder).
```
(pmp) [user@localhost folder]$ python test_embarassing_parallelization.py 
.......

Elapsed run time = 18.3232 s
```
Lastly with `mpirun`, it executes the embarassing parallelization proposed by meep.
```
(pmp) [user@localhost folder]$ mpirun -np 6 python test_embarassing_parallelization.py
.......

Elapsed run time = 18.4697 s
```

As you can see in terms of total execution time, parallelization just doesn't provide any improvement.
I should enforce that, if you substitute the simulation part of the script with any time waster computation, the total execution time reduces as expected.

**Reply by smartalecH** [reply]
Here's my system:

```
Model name:          Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz
Thread(s) per core:  1
Core(s) per socket:  8
Socket(s):           1
NUMA node(s):        1
MemTotal:       16322456 kB
```

Here are my results (i replaced 6 jobs with 8 jobs in your script):

```
> python test_embarassing_parallelization.py serial
.......
Elapsed run time = 77.4590 s
```

```
> python test_embarassing_parallelization.py 
.......
Elapsed run time = 13.4751 s
```

```
> mpirun -np 8 python test_embarassing_parallelization.py
.......
Elapsed run time = 12.8788 s
```

I see a significant speedup on my machine (not quite 8X, but close).

To answer your earlier question, I use this feature often and usually see a speedup. Especially on large clusters.

**Reply by Niccolo-Marcucci** [reply]
THANK YOU  for taking the time to test the codes!

May I ask which version of meep you run? do you use the conda install?

**Reply by smartalecH** [reply]
Glad to help :-)

I built from source, running on `master`. So a bit newer than the latest release (i.e. the conda build). But I can't imagine that's the issue, as we haven't made any real changes to `divide_parallel_processes` in several years... Plus I've used this feature myself for a couple of years now.

**Reply by Niccolo-Marcucci** [reply]
Ok I'll try to build from source then, and upadte this thread. In the cluster i got the same problem (same conda version), so hope this will solve

**Reply by FrockSaints** [reply]
Any update? I am experiencing the same issue, where I want to do a lot of independent simulations on a cluster and would prefer not to send hundreds of jobs if possible. Furthermore, there's a "big" partition that requires jobs to have a minimum number of cores, but I won't use it with this issue, as it'd be a massive waste of resources.

I could build from source, but due to inexperience and the need to have meep play nice with other libraries, including tensorflow, and deadlines approaching, it seems like a daunting task, were the process to go awry.
Furthermore, considering I've experienced slowdowns when using multiple cores for a single simulation as well, that means the parallel meep conda package doesn't provide any benefit over serial, at least in my case.

Is this happening to all conda users? Regardless, this should probably be an issue, assuming the conda parallel meep package is officially supported.

**Reply by Niccolo-Marcucci** [reply]
No I'm sorry, I gave up on trying to built from source, I tried many times and always failed. Unfortunately I don't have that kind of knowledge. (Plus, I'm pretty sure i could not do the same on the cluster, cause i have very limited permissions on it.)

On the conda side, I'm sorry to inform you that even batch parallelization won't work. I tried with a simple bash "for loop"   lunching different processes for each simulation and i got the same result i obtained [here](https://github.com/NanoComp/meep/discussions/2008#discussioncomment-2399086). 
My best "guess" is that there is some resource inside the conda meep environment that can be accessed by one process/meep-simulation at a time.

I do agree this should be an opened issue, but, since i was the only one experiencing the problem and i wasn't able to try the compiled version, i didn't insist on it.

@FrockSaints, just for the sake of clarity, did you try out also my script and obtained the same result?

**Reply by FrockSaints** [reply]
I haven't tried your script yet. I'm using serial meep on my pc and parallel meep on the cluster, so it's not so easy.
I've submitted a job with a modified version of the script that runs the 3 options in sequence and was going to wait for the result before replying, but it's taking forever to even start despite having specified only 6 cores and 5 minutes of wall time, and there being 140+ cores sitting idle for hours... Severely worrying considering how much I'll be depending on the cluster in the near future... I'll update when I get back those results.

But I had observed a similar behaviour in my own tests, except mine were about hour long simulations and the times were closer together (can't remember exact values).
When you say launching different processes for each simulation also doesn't work, do you mean individual jobs on the cluster? It's using slurm and we can specify number of cores, wall time, max memory, number of nodes, etc... I figured that could maybe help isolate the different simulations. I've actually been meaning to test if this issue happens when running 2 simulations on the same node (but different cores and jobs) versus running them on different nodes, cores and jobs, but didn't get around to it yet.

**Reply by Niccolo-Marcucci** [reply]
> When you say launching different processes for each simulation also doesn't work, do you mean individual jobs on the cluster?

Yes I do. I didn't directly test on the cluster lunching a different job for each simulation (or maybe I did and I just don't recall it), but so far everything else I tried in my personal computer worked the same in the cluster, so I'm not expecting a different behaviour. 
If you check the latest version, I updated the script so to use a bash argument as index of the for-loop. What I'm saying is that if you run the script like this:
```
for i in 0 1 2 3 4 5
do 
python test_embarassing_parallelization.py bash $i & 
done
wait
echo "done"
```
You will see all the simulations running in parallel, and each simulation taking about 10 times longer than what a single one takes when run alone. 
If this behaviour happens in this conditions, I don't see why it wouldn't using slurm. Modifying those lines for lunching sbatch is fairly straight forward, but if your waiting time is so long in your cluster, maybe those processes will end up not running simultaneously. 

P.S. By the way, if your wait time is so long for such a small process there must be something wrong, try setting also the memory usage with `#SBATCH --mem-per-cpu=100M` and maybe avoid setting the number of nodes, while setting only `#SBATCH --ntasks=6`, or maybe try to request a specific node that is idle with `#SBATCH --nodelist=example-1-14`. Good luck with that.

**Reply by FrockSaints** [reply]
Apparently, the issue was partly due to very large job requests that were on hold due to lack of resources, and were placing everyone else on hold. The cluster seems pretty much unusable right now, so I decided to build from source on my own laptop.
I followed these instructions: https://meep.readthedocs.io/en/latest/Build_From_Source/#building-from-source
After installing Ubuntu 18.04 in WSL, I used the automated script for building meep but added the --enable-single flag for single precision. `autogensh --with-mpi --with-openmp --enable-single PYTHON=python3`
Also had to add the following lines to .profile:
```
export PYTHONPATH=/usr/local/lib/python3.6/dist-packages
export PYTHONPATH=/home/frock/meep/lib/python3.6/site-packages/
```

And indeed, building from source I'm getting a speedup of parallel relative to serial. 
However this did not happen when I ran my own (larger) script, so I did some more tests with your script:
```
resolution 20:
serial: 14.3446 s
pool: 5.4757 s
mpirun: 5.3782 s
serial mpi: 10.6924 s
2 threads: 12.6818 s

resolution 25:
serial: 21.6924 s
pool: 15.4219 s
mpirun: 14.9458 s
serial mpi: 15.0225 s
2 threads: 16.9116 s

resolution 30:
serial: 31.3173 s
pool: 29.7953 s
mpirun: 31.5166 s
serial mpi: 20.5426 s
2 threads: 20.8791 s

resolution 40:
serial: 63.6589 s
pool: 76.8891 s
mpirun: 76.9603 s 
serial mpi: 37.0731 s
2 threads: 35.1372 s
```

For "serial mpi" I changed these 2 lines and ran with mpirun: `if True: #len(sys.argv) > 1:`, `if True: #sys.argv[1] == "serial":`.
This does the 6 simulations just as serial but each one with 6 cores.
For "2 threads", it's the same as "serial mpi" but with `env OMP_NUM_THREADS=2` before mpirun.

It seems the usual parallelization is getting better with bigger simulations while embarassingly parallel ones are quickly getting worse, to the point of unusability.

Granted, these are still small simulations and not a lot of tests, so it's hard to say anything for sure, but something is definitely wrong...

**Reply by FrockSaints** [reply]
@stevengj As requested (#2223, sorry for the redundancy), I opened a few terminals and ran some more tests for the serial case, both cases totalling 6 runs over all the terminals, for easier comparison with the tests above.
```
resolution 25:
2 terminals x 3 runs each: 12.7 s
6 terminals x 1 run each: 12.4 s

resolution 40:
2 terminals x 3 runs each: 54.3 s
6 terminals x 1 run each: 74.7 s
```

It's still not as effective as the usual, non embarrassing parallelization.

Here's some information regarding the memory:
```
C:\Users\USER>wmic MemoryChip get MemoryType, Name, Capacity, Configuredclockspeed, DeviceLocator, FormFactor, Manufacturer, Serialnumber, Speed
Capacity     ConfiguredClockSpeed  DeviceLocator   FormFactor  Manufacturer  MemoryType  Name             SerialNumber  Speed
17179869184  2667                  ChannelB-DIMM0  12          Kingston      0           Physical Memory  E08EA6B1      2667
```
The processor is an Intel i7-9750H and its memory bandwidth is supposedly 42GB/s.
Tried a benchmark (https://ram.userbenchmark.com/Software) and the RAM was handling 16GB/s.
Is this really so low, especially for such a small simulation?

One thing that comes to mind as a potential cause of this issue is that I'm using WSL. Maybe virtualization, or some types of virtualization, somehow mess with meep's ability to parallelize...?
I could try dual-booting if necessary, though it's a bit of a hassle, so if you've got more ideas for tests to try to pin down the cause, I'll gladly assist.
